### DISCUSSION TITLE
IMFContentProtectionManager usage

### DISCUSSION THREAD

**Discussion by TetrisSQC** [FactoryXCode/MfPack#26] [Ideas]
Hi, did someone already wrote a small sample using DRM? Otherwise I will try to build one.

Christian

**Comment by FactoryXCode** [comment]
Hi Christian.
I don't think there's a sample in the pipeline.
So, a sample from you would be appreciated.
Tony.

**Comment by TetrisSQC** [comment]
[FMX.MediaFoundation.pas.zip](https://github.com/FactoryXCode/MfPack/files/8169398/FMX.MediaFoundation.pas.zip)

Okay this version is a translation of this example:
https://github.com/GoshaDE/SuperMFLib/blob/master/samples/ProtectedPlayback/Player.cs

Further on It is a TControl derived component for Firemonkey. You can put all FMX components on it and they will be rendered over the video stream. It allows you to paint an OSD without any performance loss. Anyhow I did not got the DRM class working. It seem to be that it won't be called at all. Maybe somebody can have a look on it?

**Reply by FactoryXCode** [reply]
I don't think this the compleet sample is it?

**Reply by TetrisSQC** [reply]
To complete the sample, you just have to create an empty project and add a:
 FPlayer := TMediaEngine.create(self);
 FPlayer.parent := self;
 FPlayer.Align := TFMXAlignLayout.Client;
 FPlayer.Filename := 'MyFile';
 FPlayer.Play;

I wrote a player it should be found in the post below.

**Comment by TetrisSQC** [comment]
[fmx player.zip](https://github.com/FactoryXCode/MfPack/files/8192309/fmx.player.zip)
The attached zip contain a player with an simple overlay running using Firemonkey.
![player](https://user-images.githubusercontent.com/11291557/156918130-e48cfd8f-e5a9-46ba-b296-49f22edcf2eb.PNG)

I tried to keep it as simple as possible. Anyhow DRM does not seem to work. If the application crash it is most likely this line:
(FMediaEngine as IMFMediaEngineProtectedContent).SetContentProtectionManager(FContentProtectionManager);
It happen from time to time, but I do not see what I did wrong while implementing the ContentProtectionManager interface.

Christian

**Reply by FactoryXCode** [reply]
Ok , I tested the sample.
This what's changed & fixed:

Reformatted the code.
Fixed memory leaks
Fixed wrong project code (Form3 issue)
Tested the sample and no issues you mentioned happens.

Open issues:
  The alignments of the layEPGCenter etc. labels are not correct when not full screen, in full screen mode they appears to be 
  aligned  correctly.

Conclusion: FMediaEngine as IMFMediaEngineProtectedContent).SetContentProtectionManager(FContentProtectionManager; works as expected. (no return code different from S_OK) and is called on initialization without errors.

The corrected sample code has been put [here](https://github.com/FactoryXCode/MfPack/tree/Master/MfPack/SandBox/ProtectedPlayBack).

Maybe the code you wrote for your project differs from the code of this sample?

Regards, Tony.

**Comment by TetrisSQC** [comment]
Thanks, 
yes I did not checked possible leaks at this point. The epg was just an example - the whole scaling is done by the TScaledLayout class. Once I got DRM working I will build a more useful example. 
I think I found the reason for my crashes. The main development here is done on a Mac using Parallels with an Windows 10 VM and it crashes in the graphic driver.
Besides this the DRM interface should be called if you for example want to play a stream like this one:
http://profficialsite.origin.mediaservices.windows.net/c51358ea-9a5e-4322-8951-897d640fdfd7/tearsofsteel_4k.ism/manifest(format=mpd-time-csf)

Not sure if the MF_MEDIA_ENGINE_CONTENT_PROTECTION_MANAGER should be used instead.

Christian

**Reply by FactoryXCode** [reply]
> Not sure if the MF_MEDIA_ENGINE_CONTENT_PROTECTION_MANAGER should be used instead.

According to [this article](https://docs.microsoft.com/en-us/windows/win32/medfound/how-to-play-protected-media-files), you have. 

Also: function MFCreatePMPMediaSession() is not implemented. To play DRM protected media, you have to initiate a PMP (Protected Media Path). 

Usually, as I mentioned before, you should detect DRM protection first. However Microsoft say's this is optional, but make is easier to determine if a PMP or normal session would be needed.

A sample how to :


```
// If this function returns TRUE, it means the PMP is required for this presentation.
// Call MFCreatePMPMediaSession to create the PMP session object.
//
// If the function returns S_FALSE, you can use the unprotected pipeline.
// Call MFCreateMediaSession to create the regular Media Session object.
//
// Internally, this function checks whether any of the stream descriptors in the
// presentation have the MF_SD_PROTECTED attribute with the value TRUE.
function TContentProtectionManager.RequireDRMRights(): Boolean;
var
  hr: HResult;
  pPresentationDescriptor: IMFPresentationDescriptor;
  apStreamDescriptors: IMFStreamDescriptor;

begin

  hr := MFCreatePresentationDescriptor(0,
                                       @apStreamDescriptors,
                                       pPresentationDescriptor);



  hr :=  MFRequireProtectedEnvironment(pPresentationDescriptor);
  Result :=  Succeeded(hr);
end;
```

Regards Tony

**Comment by TetrisSQC** [comment]
Thanks, 

but how do I use MFCreatePMPMediaSession together with the IMFMediaEngine? I tried play dash streams using the MFCreateMediaSession based projects instead of the IMFMediaEngine, but they fail playing these king of streams. They work with mp4 files, but everything else is not working.

Regards
Christian

**Reply by FactoryXCode** [reply]
OoPs, I was referring to the session methods, Sorry.  
As what I known is, that not all protected media can be handled by those interfaces. There is a default way to handle DRM protected media and that is the default implementation, but in some cases you need to have a decryption key (2 keys actually) those are a 'kid' and 'key'.
Both are defined, for example, as kid = eb676abbcb345e96bbcf616630f1a3da and key 100b6c20940f779a4589152b57d2dacb to decrypt a  DRM protected mediastream. So, if you are lucky a mediastream can be decoded within a PMP and else you need the decoder keys provided by the owner of the content. See [here](https://github.com/FactoryXCode/MfPack/tree/Master/MfPack/SandBox/MFProtectedPlayBack/DRM%20Protected%20Files) for a media file that needs those key's. Maybe [this](https://forum.videohelp.com/threads/401172-Everyone-can-help-me-fix-error-on-mp4decrypt) explains why not every DRM protected media can be played or handled.
I tried to play this file in VLC and MediaPlayer and both are not capable of playing this file. 

Regards Tony.

**Reply by TetrisSQC** [reply]
Hi, yes I know I do have a certificate and the key url where the DRM method gets the key. This works fine with Widevine and Fairplay with the APIs provided by google resp. apple. The project I work on also requires a windows build so I got stuck in the very poorly documented PlayReady. I thought in theory this would work similar. An async callback asks for the certificate and then the content key for decoding.

<img width="1810" alt="Bildschirmfoto 2022-03-12 um 09 47 41" src="https://user-images.githubusercontent.com/11291557/158011005-c0891605-688d-4db7-9606-39e2c7ca15dc.png">

Christian

**Reply by FactoryXCode** [reply]
Maybe the IMFMediaEngineNeedKeyNotify callback interface is also needed to notify key request data?

If found [this](https://github.com/microsoft/media-foundation/tree/master/samples/MediaEngineEMEUWPSample) a C++ sample to use PlayReady with the MediaEngine API.

Regards, Tony.

**Reply by TetrisSQC** [reply]
Hmm I'm not sure, I found an implementation in chrome here:
https://chromium.googlesource.com/chromium/src/media/+/refs/heads/main/renderers/win/media_foundation_renderer.cc
The only difference between our implementation and this one seem to be the MF_MEDIA_ENGINE_OPM_HWND handle, but I am not quite familiar with WMF. 

Christian

**Reply by FactoryXCode** [reply]
> MF_MEDIA_ENGINE_OPM_HWND 

This wil put the MediaEngine in rendering mode and is a required attribute according to [this article](https://docs.microsoft.com/en-us/windows/win32/api/mfmediaengine/nf-mfmediaengine-imfmediaengineclassfactory-createinstance#initialization-attributes).

> WMF
This is just a container and a Windows native format. For rendering there is nothing required like MP4, AVI and all other formats supported by Media Foundation. What wonders me is why the callback is not working correctly.

Regards, Tony.

**Reply by FactoryXCode** [reply]
Also what wonders me is that (FMediaEngine as IMFMediaEngineProtectedContent).SetOPMWindow(FWnd) returns E_NOTIMPL. While this function has to be called just before the command Play()

**Reply by TetrisSQC** [reply]
Yes, I stumbled over this also yesterday. You could use hr := li_Attributes.SetUINT64(MF_MEDIA_ENGINE_OPM_HWND, FWnd);
but this is probably the same and it seem to be that this is not implemented in the Mediafoundation. My second thought was to use MFCreatePMPMediaSession but IMFSourceResolver fails when it comes to Dash Urls. I also did not found a explanation how to handle Dashs in this mode in general. CreateObjectFromURL returns NS_E_SERVER_NOT_FOUND on every dash or hls stream I tested.

**Reply by FactoryXCode** [reply]
> IMFSourceResolver fails when it comes to Dash Urls

The parameter pwszURL must be a local URL or URL. The function will not parse Object-URL (blob) data. 

Regards, Tony.

**Reply by TetrisSQC** [reply]
These days I found time to work on this DRM related code a bit. It seem to be that you have to create a OPM window.

```
if not GetClassInfoEx(SysInit.HInstance,
    PChar('VirtualMediaFoundationCdmVideoWindow'), WindowClassEx) then
  begin
    FillChar(WindowClassEx, SizeOf(WindowClassEx), 0);
    WindowClassEx.cbSize := SizeOf(TWndClassEx);
    WindowClassEx.Style := CS_OWNDC;
    WindowClassEx.lpfnWndProc := @DefWindowProc;
    WindowClassEx.cbClsExtra := 0;
    WindowClassEx.cbWndExtra := 0;
    WindowClassEx.HInstance := SysInit.HInstance;
    WindowClassEx.hCursor := LoadCursorW(0, PChar(IDC_ARROW));
    WindowClassEx.hbrBackground := GetStockObject(BLACK_BRUSH);
    WindowClassEx.lpszMenuName := nil;
    WindowClassEx.lpszClassName :=
      PChar('VirtualMediaFoundationCdmVideoWindow');
    if WinApi.Windows.RegisterClassEx(WindowClassEx) = 0 then
      RaiseLastOSError;
  end;

  FVirtualWindow := CreateWindowEx(WS_EX_NOPARENTNOTIFY or WS_EX_LAYERED or
    WS_EX_TRANSPARENT or WS_EX_NOREDIRECTIONBITMAP, WindowClassEx.lpszClassName,
    '', WS_POPUP or WS_DISABLED or WS_CLIPSIBLINGS, 0, 0, 1, 1, 0, 0, 0, nil);

  li_Attributes.SetUnknown(MF_MEDIA_ENGINE_CALLBACK, self);
  li_Attributes.SetUINT64(MF_MEDIA_ENGINE_PLAYBACK_HWND, FWnd);
  li_Attributes.SetUINT64(MF_MEDIA_ENGINE_OPM_HWND, FVirtualWindow);
  li_Attributes.SetUINT32(MF_MEDIA_ENGINE_CONTENT_PROTECTION_FLAGS,
    MF_MEDIA_ENGINE_ENABLE_PROTECTED_CONTENT);
  li_Attributes.SetUnknown(MF_MEDIA_ENGINE_CONTENT_PROTECTION_MANAGER,
    FContentProtectionManager);
```

Anyhow the difference now is that if I add the protection manager the application crashes while loading drm encrypted content. I wonder why all the examples from Microsoft are UWP apps and it seem to be that they never really tested play ready with a "native" 32/64bit application.

**Reply by FactoryXCode** [reply]
Sorry to be a "bit" late...

I think FContentProtectionManager needs an implemented callback from the browser or object that is processing the URL. Microsoft has a sample to do this, but is based on Internet Explorer...

**Comment by FactoryXCode** [comment]
Any one else have an idea?

**Comment by FactoryXCode** [comment]
Ok, I managed to solve the issues with protected paths.
As I replied, the sample uses Internet Explorer automation COM server. I added a Powershell script, to check if a windows version supports it.
With my latest Windows 11 version (25H2, 26200.9457), the Internet Explorer automation COM server is still there.
We brought out the new sample MfProtectedPlayback (the Delphi translation of Microsoft's sample "ProtectedPlayback".
See: [MfProtectedPlayback](https://github.com/FactoryXCode/MfPack/tree/Master/MfPack/Samples/MfProtectedPlayback)
There is one thing though: Files like FLAC are not supported.
FLAC itself is supported by Media Foundation, but standalone .flac files on a machine could be
not accepted inside the protected PMP process.

Regards, Tony.
